pennylane/transforms/get_unitary_matrix.py

```diff
@@ -0,0 +1,119 @@
+from difflib import SequenceMatcher
+from functools import wraps, reduce
+
+from pennylane.wires import Wires
+from pennylane.tape import QuantumTape, get_active_tape
+import numpy as np
+
+
+def get_unitary_matrix(fn, wire_order):
+    """Given a quantum circuit and a list of wire ordering, construct the matrix representation"""
+
+    wire_order = Wires(wire_order)
+    n_wires = len(wire_order)
+
+    @wraps(fn)
+    def wrapper(*args, **kwargs):
+        ######### NOTE BELOW
+        active_tape = get_active_tape()
+
+        if active_tape is not None:
+            with active_tape.stop_recording(), QuantumTape() as tape:
+                fn(*args, **kwargs)
+        else:
+            # Not within a queuing context
+            with QuantumTape() as tape:
+                fn(*args, **kwargs)
+
+        if not tape.operations:
+            # we called op.expand(): get the outputted tape
+            tape = fn(*args, **kwargs)
+        #### NOTE took the above from qml.adjoint. Appropriate here?
+
+        # initialize the unitary matrix
+        unitary_matrix = np.eye(2 ** n_wires)
+
+        for op in tape.operations:
+
+            # get the matrix for the operator op
+            op_matrix = _get_op_matrix(op, wire_order, [np.eye(2 ** n_wires)])
+            # add the single operator matrix to the full unitary matrix
+            unitary_matrix = np.dot(op_matrix, unitary_matrix)
+
+        return unitary_matrix
+
+    return wrapper
+
+
+def _get_op_matrix(op, wire_order, SWAP_list):
+    """Called by get_unitary_matrix(fn, wire_order) to construct the matrix representation for an operator"""
+    # check if the operator wires match the wire order
+    s = SequenceMatcher(lambda x: x == " ", op.wires.tolist(), wire_order.tolist())
```

Comment:
For this part, I recommend looking into how PennyLane performs statevector simulation, as internally it performs this matrix construction and application; it's just buried in the devices. For example, in `devices/default_qubit.py`, there is a function called `_apply_unitary()`. It is used to both construct and apply unitaries to a statevector, but you can see there how the wires get mapped and permuted and maybe adapt these to the transform.

Reply:
Ahh, yeah I was wondering how the simulations were performed numerically if the unitary wasn't already calculated :P... I tried to test something like `_applied_unitary()` now but tensordot gave me trouble. But since the tensordot is used to calculate the state, and we don't want a state, I guess I shouldn't even need it, but it seems a bit trickty to figure out... But if it is doable, then swaps won't even be needed?

Reply:
In the meantime, while thinking and searching about how this could be done, I found a neat way to generate the swaps in 1 line of code...

Reply:
Wow I can't believe I actually made something that seems to work!? Honestly I don't even know exactly how it works, it's just based on code from `_apply_unitary()` in `devices/default_qubit.py` and just some manual testing and fiddling and looking at what came out. I have a feeling the matrix `I` could be removed if the tensordot is adjusted properly, but I'm still not fully clear on how tensordot works.